### PR TITLE
Add k34 keyboard

### PR DESCRIPTION
**Pull request by wongjingping**
Add a keyboard for a custom k34 keyboard that I designed recently.

## Description

This creates the folder structure and required files to build and flash the firmware for standard pro-micro MCU's. Instructions are in the readme.md

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

Added a keyboard for the k34. The only thing missing is the layout in the `info.json` which is quite complicated for this ortho layout and have left out for the time being.

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/k34/rules.mk [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+CONSOLE_ENABLE = no         # Console for debug
+COMMAND_ENABLE = no         # Commands for debug and configuration
+# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
+SLEEP_LED_ENABLE = no       # Breathing sleep LED during USB suspend
+# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
+NKRO_ENABLE = no            # USB Nkey Rollover
+BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
+RGBLIGHT_ENABLE = no        # Enable keyboard RGB underglow
+AUDIO_ENABLE = no           # Audio output
+COMBO_ENABLE = yes          # enable combo
```
Comment (zvecr) [inline keyboards/k34/rules.mk:22]: Should only be enabled at the keymap level
```suggestion
```

File: keyboards/k34/config.h [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+#define DEBOUNCE 5
+
+// use 1 combo
+#define COMBO_COUNT 1
+#define COMBO_TERM 300
+
+// mod taps for home row mods
+#define TAPPING_TERM_PER_KEY
+#define IGNORE_MOD_TAP_INTERRUPT
+
```
Comment (zvecr) [inline keyboards/k34/config.h:46]: Should only be enabled at the keymap level
```suggestion
```

File: keyboards/k34/k34.h [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+/* This is a shortcut to help you visually see your layout.
+ *
+ * The first section contains all of the arguments representing the physical
+ * layout of the board and position of the keys.
+ *
+ * The second converts the arguments into a two-dimensional array which
+ * represents the switch matrix.
+ */
+#define LAYOUT_K34( \
```
Comment (zvecr) [inline keyboards/k34/k34.h:16]: Not a valid layout macro name
```suggestion
#define LAYOUT( \
```

File: keyboards/k34/keymaps/default/keymap.c [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+#define G_RBRC  G(KC_RBRC)
+#define A_SLSH  A(KC_SLSH)
+
+/* combos */
+const uint16_t PROGMEM jk_combo[] = {KC_J, KC_K, COMBO_END};
+combo_t key_combos[COMBO_COUNT] = {COMBO(jk_combo, KC_ESC)};
+
+const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
+    /* Base */
+  [_BASE] = LAYOUT_K34(
```
Comment (zvecr) [inline keyboards/k34/keymaps/default/keymap.c:40]:
```suggestion
  [_BASE] = LAYOUT(
```

File: keyboards/k34/keymaps/default/keymap.c [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    KC_Q,     KC_W,     KC_E,     KC_R,      KC_T,               KC_Y,     KC_U,     KC_I,     KC_O,     KC_P,
+    G_A,      KC_S,     KC_D,     KC_F,      KC_G,               KC_H,     KC_J,     KC_K,     KC_L,     G_SC,
+    S_Z,      A_X,      C_C,      KC_V,      KC_B,               KC_N,     KC_M,     C_COMM,   A_DOT,    S_SLSH,
+				  L1,        KC_SPC,             KC_SPC,   L2
+  ),
+
+
+  [_L1] = LAYOUT_K34(
```
Comment (zvecr) [inline keyboards/k34/keymaps/default/keymap.c:48]:
```suggestion
  [_L1] = LAYOUT(
```

File: keyboards/k34/keymaps/default/keymap.c [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+    KC_EXLM,  KC_AT,    KC_HASH,  KC_DLR,   KC_PERC,            KC_CIRC,  KC_AMPR,  KC_ASTR,  KC_MINS,  KC_EQL,
+    KC_LPRN,  KC_RPRN,  KC_LCBR,  KC_RCBR,  KC_GRV,             KC_QUOT,  KC_LBRC,  KC_RBRC,  KC_BSLS,  KC_PIPE,
+    _______,  _______,  _______,  KC_TAB,   _______,            KC_MUTE,  KC_VOLD,  KC_VOLU,  KC_MPLY,  KC_MNXT,
+				  _______,  _______,            _______,  _______
+  ),
+
+  [_L2] = LAYOUT_K34(
```
Comment (zvecr) [inline keyboards/k34/keymaps/default/keymap.c:55]:
```suggestion
  [_L2] = LAYOUT(
```

**Review by zvecr — CHANGES_REQUESTED** [review]
Also, `info.json` missing

File: keyboards/k34/readme.md [hunk 7]
```diff
@@ -0,0 +1,19 @@
+# k34
+
+![k34](https://github.com/wongjingping/k34/raw/main/docs/k34.jpg?raw=true)
+
+*A short description of the keyboard/project*
```
Comment (fauxpark) [inline keyboards/k34/readme.md:5]: Please fill this out.

File: keyboards/k34/readme.md [hunk 8]
```diff
@@ -0,0 +1,19 @@
+# k34
+
+![k34](https://github.com/wongjingping/k34/raw/main/docs/k34.jpg?raw=true)
```
Comment (fauxpark) [inline keyboards/k34/readme.md:3]: Please rehost this on Imgur.

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/k34/readme.md [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+
+*A short description of the keyboard/project*
+
+* Keyboard Maintainer: [wongjingping](https://github.com/wongjingping)
+All hardware supported can be found in the [k34 repository](https://github.com/wongjingping/k34)
+
+Make example for this keyboard (after setting up your build environment):
```
Comment (drashna) [inline keyboards/k34/readme.md:10]: Could you add instructions on how to enter the bootloader, to flash new firmware? 

Eg: 
```markup
## Bootloader

Enter the bootloader in 3 ways:

* **Bootmagic reset**: Hold down the key at (0,0) in the matrix (usually the top left key or Escape) and plug in the keyboard
* **Physical reset button**: Briefly press the button on the back of the PCB - some may have pads you must short instead
* **Keycode in layout**: Press the key mapped to `RESET` if it is available
```

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/k34/rules.mk (hunk 1 above)
Comment (wongjingping) [reply keyboards/k34/rules.mk]: Done

File: keyboards/k34/config.h (hunk 2 above)
Comment (wongjingping) [reply keyboards/k34/config.h]: Done

File: keyboards/k34/k34.h (hunk 3 above)
Comment (wongjingping) [reply keyboards/k34/k34.h]: Done

File: keyboards/k34/keymaps/default/keymap.c (hunk 4 above)
Comment (wongjingping) [reply keyboards/k34/keymaps/default/keymap.c]: Done

File: keyboards/k34/keymaps/default/keymap.c (hunk 5 above)
Comment (wongjingping) [reply keyboards/k34/keymaps/default/keymap.c]: Done

File: keyboards/k34/keymaps/default/keymap.c (hunk 6 above)
Comment (wongjingping) [reply keyboards/k34/keymaps/default/keymap.c]: Done

File: keyboards/k34/readme.md (hunk 7 above)
Comment (wongjingping) [reply keyboards/k34/readme.md]: Done

File: keyboards/k34/readme.md (hunk 8 above)
Comment (wongjingping) [reply keyboards/k34/readme.md]: Done

File: keyboards/k34/readme.md (hunk 9 above)
Comment (wongjingping) [reply keyboards/k34/readme.md:16]: Done

File: keyboards/k34/keymaps/default/config.h [hunk 10]
```diff
@@ -0,0 +1,9 @@
+
```
Comment (drashna) [inline keyboards/k34/keymaps/default/config.h:1]: Could you add a GPL2+ compatible license header here? 

For instance:
```c
// Copyright 2021 Your Name (@yourgithub)
// SPDX-License-Identifier: GPL-2.0-or-later
```

Or 

https://github.com/qmk/qmk_firmware/blob/6b299fab6e6a3143c061b9ea005253361fcd524b/data/templates/base/keyboard.c#L1-L15

Also, this need as an include guard, eg `#pragma once`

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/k34/readme.md [hunk 11]
```diff
@@ -0,0 +1,25 @@
+# k34
+
+![k34](https://imgur.com/gallery/ZELj0gR)
```
Comment (fauxpark) [inline keyboards/k34/readme.md:3]:
```suggestion
![k34](https://i.imgur.com/tarkFIp.jpeg)
```
Needs to be a direct link.

**Review by fauxpark — CHANGES_REQUESTED** [review]

**Review by fauxpark — CHANGES_REQUESTED** [review]
Also missing an info.json.

File: keyboards/k34/info.json [hunk 12]
```diff
@@ -0,0 +1,181 @@
...
+    "keyboard_name": "k34", 
+    "url": "https://github.com/wongjingping/k34", 
+    "maintainer": "qmk", 
+    "layouts": {
+        "LAYOUT": {
+            "layout":[
+                {
+                    "label":"E",
+                    "x":2,
+                    "y":-0.1
```
Comment (fauxpark) [inline keyboards/k34/info.json:11]: Key coordinates should be positive numbers. Please check this file is correct by pressing Ctrl+Shift+I in the Configurator.

**Review by drashna — APPROVED** [review]

File: keyboards/k34/info.json (hunk 12 above)
Comment (wongjingping) [reply keyboards/k34/info.json:11]: This is due to the unconventional layout of the keys it seems - let me know if this is a blocker that I should resolve.

**Comment by wongjingping** [comment]
> Also, `info.json` missing

Added an info.json. Given that this uses an unconventional layout, the positions are a little "abnormal".

File: keyboards/k34/info.json (hunk 12 above)
Comment (fauxpark) [reply keyboards/k34/info.json:11]: Yes, it doesn't look correct. You need to redo it without the rotation on the keys (I can see one of them has a rounding error) and place the sides together.
image

Use the Configurator to test the file, by pressing Ctrl+Shift+I. You can also make sure the keys are in the same order as the layout macro by turning on Fast Input.

File: keyboards/k34/rules.mk [hunk 13]
```diff
@@ -0,0 +1,21 @@
...
+# Build Options
+#   change yes to no to disable
+#
+BOOTMAGIC_ENABLE = yes      # Enable Bootmagic Lite
+MOUSEKEY_ENABLE = no        # Mouse keys
+EXTRAKEY_ENABLE = yes       # Audio control and System control
+CONSOLE_ENABLE = no         # Console for debug
+COMMAND_ENABLE = no         # Commands for debug and configuration
+# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
+SLEEP_LED_ENABLE = no       # Breathing sleep LED during USB suspend
```
Comment (fauxpark) [inline keyboards/k34/rules.mk:16]:
```suggestion
```

File: keyboards/k34/readme.md [hunk 14]
```diff
@@ -0,0 +1,25 @@
+# k34
+
+![k34](https://i.imgur.com/tarkFIp.jpeg)
```
Comment (fauxpark) [inline keyboards/k34/readme.md:3]:
```suggestion
![k34](https://i.imgur.com/tarkFIph.jpeg)
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/k34/info.json (hunk 12 above)
Comment (wongjingping) [reply keyboards/k34/info.json]: Screen Shot 2021-12-05 at 10 20 35 PM
Hmm I'm not sure why but it seems like the negative number is still there after I import it with ctrl-shift-I. Any tips for how to fix that?

Comment (fauxpark) [reply keyboards/k34/info.json]: Assuming you generated it from the KLE, you have to fix that.

Comment (wongjingping) [reply keyboards/k34/info.json]: changed it to 0 instead of a negative number - would that work?

File: keyboards/k34/rules.mk [hunk 15]
```diff
@@ -0,0 +1,19 @@
...
+# Build Options
+#   change yes to no to disable
+#
+BOOTMAGIC_ENABLE = yes      # Enable Bootmagic Lite
+MOUSEKEY_ENABLE = no        # Mouse keys
+EXTRAKEY_ENABLE = yes       # Audio control and System control
+CONSOLE_ENABLE = no         # Console for debug
+COMMAND_ENABLE = no         # Commands for debug and configuration
+# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
+NKRO_ENABLE = no            # USB Nkey Rollover
```
Comment (fauxpark) [inline keyboards/k34/rules.mk:16]:
```suggestion
NKRO_ENABLE = no            # Enable N-Key Rollover
```

File: keyboards/k34/info.json [hunk 16]
```diff
@@ -0,0 +1,10 @@
...
+    "maintainer": "qmk", 
+    "layouts": {
+        "LAYOUT": {
+            "layout": [{"label":"E", "x":2, "y":0.0}, {"label":"W", "x":1, "y":0.25}, {"label":"R", "x":3, "y":0.25}, {"label":"Q", "x":0, "y":0.50}, {"label":"T", "x":4, "y":0.6}, {"label":"D", "x":2, "y":0.9}, {"label":"S", "x":1, "y":1.25}, {"label":"F", "x":3, "y":1.25}, {"label":"A", "x":0, "y":1.5}, {"label":"G", "x":4, "y":1.6}, {"label":"C", "x":2, "y":1.9}, {"label":"X", "x":1, "y":2.25}, {"label":"V", "x":3, "y":2.25}, {"label":"Z", "x":0, "y":2.5}, {"label":"B", "x":4, "y":2.6}, {"label":"Shift", "x":3, "y":3.25}, {"label":"Bksp", "x":4, "y":3.6}, {"label":"I", "x":2, "y":4.4}, {"label":"U", "x":1, "y":4.75}, {"label":"O", "x":3, "y":4.75}, {"label":"P", "x":4, "y":5.0}, {"label":"Y", "x":0, "y":5.1}, {"label":"K", "x":2, "y":5.4}, {"label":"J", "x":1, "y":5.75}, {"label":"L", "x":3, "y":5.75}, {"label":":", "x":4, "y":6.0}, {"label":"H", "x":0, "y":6.1}, {"label":"<", "x":2, "y":6.4}, {"label":"M", "x":1, "y":6.75}, {"label":">", "x":3, "y":6.75}, {"label":"?", "x":4, "y":7.0}, {"label":"N", "x":0, "y":7.1}, {"label":"fn", "x":1, "y":7.75}, {"label":"Space", "x":0, "y":8.1}]
```
Comment (fauxpark) [inline keyboards/k34/info.json:7]: Still not correct. The two sides need to be next to each other, and the key ordering needs to match the layout macro. Use the Configurator to test this, press Ctrl+Shift+I and turn on fast input.

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/k34/info.json (hunk 16 above)
Comment (wongjingping) [reply keyboards/k34/info.json]: Think I might have got it; was referencing some other similar keyboards but please let me know if I'm still missing something!

File: keyboards/k34/rules.mk [hunk 17]
```diff
@@ -0,0 +1,18 @@
...
+
+# Build Options
+#   change yes to no to disable
+#
+BOOTMAGIC_ENABLE = yes      # Enable Bootmagic Lite
+MOUSEKEY_ENABLE = no        # Mouse keys
+EXTRAKEY_ENABLE = yes       # Audio control and System control
+CONSOLE_ENABLE = no         # Console for debug
+COMMAND_ENABLE = no         # Commands for debug and configuration
+NKRO_ENABLE = no            # USB Nkey Rollover
```
Comment (fauxpark) [inline keyboards/k34/rules.mk:15]:
```suggestion
NKRO_ENABLE = no            # Enable N-Key Rollover
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

**Review by drashna — APPROVED** [review]

**Review by fauxpark — APPROVED** [review]

**Comment by wongjingping** [comment]
> Also, `info.json` missing

added info.json

**Pull request merged**